chore: implement FSM for machines and machine worker on new provisioner task

nvinuesa opened the pull request:
*This patch is the first of a series that will eventually replace the provisioner_task current implementation. This series follows the spec in https://docs.google.com/document/d/1rJelL5N5V54MVjA74H2-r9YlXWwfaT3cly1KjDMWmv4.*

This patch creates a new package (`provisionertaskbis`) that will serve as the scaffold for the new implementation and will live (unwired) along with the current implementation to avoid disrupting Juju until the final wiring.

So the first thing (first commit) is to implement a finite state machine for machine states that will allow for explicit handling of the machine's lifecycle:

<img width="933" height="697" alt="image" src="https://github.com/user-attachments/assets/036f4317-bbd2-48b3-9a0c-136c3f007b55" />


| State           | Description                                                    |
|-----------------|----------------------------------------------------------------|
| `Pending`       | Machine exists, no instance ID, awaiting provisioning          |
| `RequestingZone`| Waiting for AZ Coordinator to assign a zone                    |
| `Provisioning`  | StartInstance + SetInstanceInfo calls in flight                |
| `Running`       | Instance created & registered, life=Alive                      |
| `Stopping`      | StopInstances call in flight                                   |
| `Removing`      | Removing machine record (no instance to stop, or already stopped) |
| `Done`          | Terminal state, worker exits cleanly                           |

Also, as a second step, we provide a (scaffolding with limited functionality) of the machine worker. Machine worker fits in the new architecture as shown:
```
┌─────────────────────────────────────────────────────────────────────────────────┐
│                           Provisioner Task (Main Loop)                           │
│                                                                                  │
│  ┌─────────────┐         ┌──────────────────┐    ┌──────────────────────────┐   │
│  │   Machine   │         │                  │    │   Provider Semaphore     │   │
│  │   Watcher   │────────►│   Event Router   │    │  (resizable, token-based)│   │
│  └─────────────┘         │                  │    └──────────────────────────┘   │
│                          │  - machineChanges│                 ▲                  │
│  ┌─────────────┐         │  - AZ responses  │                 │                  │
│  │   Model     │────────►│  - AZ requests   │                 │                  │
│  │   Config    │         │    (from workers)│                 │                  │
│  └─────────────┘         └────────┬─────────┘                 │                  │
│                                   │                           │                  │
│                    ┌──────────────┼──────────────┐            │                  │
│                    │              │              │            │                  │
│                    ▼              ▼              ▼            │                  │
│             ┌───────────┐  ┌───────────┐  ┌───────────┐       │                  │
│             │  Worker   │  │  Worker   │  │  Worker   │       │                  │
│             │  Handle   │  │  Handle   │  │  Handle   │       │                  │
│             │           │  │           │  │           │       │                  │
│             │ eventsCh  │  │ eventsCh  │  │ eventsCh  │       │                  │
│             └─────┬─────┘  └─────┬─────┘  └─────┬─────┘       │                  │
│                   │              │              │              │                  │
└───────────────────┼──────────────┼──────────────┼──────────────┼──────────────────┘
                    │              │              │              │
                    ▼              ▼              ▼              │
             ┌───────────┐  ┌───────────┐  ┌───────────┐         │
             │  Machine  │  │  Machine  │  │  Machine  │         │
             │  Worker   │  │  Worker   │  │  Worker   │─────────┘
             │   (FSM)   │  │   (FSM)   │  │   (FSM)   │  (acquire/release)
             │           │  │           │  │           │
             │ machine-0 │  │ machine-1 │  │ machine-2 │
             └───────────┘  └───────────┘  └───────────┘

                    ┌──────────────────────────────────┐
                    │         AZ Coordinator           │
                    │       (catacomb-managed)         │
                    │                                  │
                    │  - Receives refresh from main    │
                    │  - Receives requests from main   │
                    │  - Sends responses to main       │
                    │  - Zone assignments              │
                    │  - Tentative allocations         │
                    │  - Failure tracking              │
                    │  - Instance presence             │
                    └──────────────────────────────────┘
```
And will be responsible of keeping each machine's state explicitly with its FSM struct.

## QA steps

Nothing wired yet, so only unit tests should pass.

nvinuesa committed:
feat: implement FSM states and transitions

FSM states:
- Pending: initial state, ready to begin provisioning.
- RequestingZone: waiting for AZ Coordinator response.
- Provisioning: executing StartInstance and SetInstanceInfo.
- Running: instance created and registered, waiting for death.
- Stopping: executing StopInstances.
- Removing: removing machine record from state.
- Done: terminal state, worker exits cleanly.

State transitions enforce valid paths through the provisioning lifecycle
including retry loops and terminal behavior.

nvinuesa committed:
feat: add events and worker request types

Add the event and request types for communication between the main loop
and machine workers:

Events (main loop -> worker):
- EventLifeChanged: machine life value changed.
- EventZoneAssigned: AZ Coordinator assigned a zone.
- EventZoneRequestFailed: zone request could not be fulfilled.

Worker requests (worker -> main loop):
- RequestZone: ask for zone assignment.
- RequestProvisionComplete: report provisioning outcome.
- RequestCancelZone: cancel pending zone request.

Also includes StartInstanceParams and StartInstanceResult types for
the broker interface.

nvinuesa committed:
feat: add service interfaces for worker dependencies

Add service interfaces that abstract the worker's external dependencies
for testability:

- MachineAPI: operations on a machine (ID, Life, InstanceID, KeepInstance,
  EnsureDead, SetStatus, SetInstanceStatus, MarkForRemoval).
- InstanceBroker: provisioning operations (StartInstance, StopInstances).
- InstanceInfoSetter: register instance information for a machine.
- ProviderSemaphore: limit concurrent provider API calls.
- Logger: logging functionality for the worker.

nvinuesa committed:
feat: implement MachineWorker with FSM-based lifecycle

Add the MachineWorker that manages a single machine's provisioning lifecycle
using the FSM. The worker receives events through a channel from the main
loop and sends requests back for coordination.

Key features:
- Event-driven architecture with single event channel.
- FSM-enforced state transitions.
- Retry logic with configurable delay and max attempts.
- Semaphore-protected provider API calls.
- Rollback on SetInstanceInfo failure (StopInstances before retry).
- Keep-instance semantics (no StopInstances when keep-instance=true).
- Clean shutdown via catacomb integration.

Edge cases handled:
- Machine dies while provisioning (queued event handled after completion).
- Retry exhaustion sets ProvisioningError and worker exits.
- Stale events (zone assigned/failed in wrong state) are ignored.
- Orphaned dead machine cleanup (no provider stop, just remove).

nvinuesa committed:
chore: change to type of FSM states to string

Replace iota-based int State with explicit string constants and simplify
String() to return the underlying string value instead of a switch.

nvinuesa commented on internal/provisionertaskbis/events.go:
```diff
@@ -0,0 +1,135 @@
+	// For RequestZone: ZoneRequestPayload.
+	// For RequestProvisionComplete: ProvisionResultPayload.
+	// For RequestCancelZone: nil (MachineID is sufficient).
+	Payload any
```
This is left as `any` for scaffolding purposes only.

nvinuesa committed:
chore: move to string-based types for event/request enums

Small change/refactor to have string event types.

SimonRichardson commented on internal/provisionertaskbis/fsm.go:
```diff
@@ -0,0 +1,139 @@
+	if !ok {
+		return false
+	}
+	return targets[target]
```
```suggestion
	_, ok := targets[target]
	return ok
```

SimonRichardson commented on internal/provisionertaskbis/machine_worker.go:
```diff
@@ -0,0 +1,598 @@
+	for {
+		select {
+		case <-w.catacomb.Dying():
+			w.stopRetryTimer()
```
Put this as `defer w.stopRetryTimer()` outside of the for loop.

SimonRichardson commented on internal/provisionertaskbis/machine_worker.go:
```diff
@@ -0,0 +1,598 @@
+	config   MachineWorkerConfig
+
+	// FSM manages state transitions.
+	fsm *FSM
```
If this bounces, we'll start at the initial state. Something to consider.

SimonRichardson commented on internal/provisionertaskbis/machine_worker.go:
```diff
@@ -0,0 +1,598 @@
+	Logger Logger
+
+	// EventsChan receives events from the main loop.
+	EventsChan <-chan MachineEvent
```
I wonder if it's better for each machine worker to have their own channel, and we publish to that as a method on the worker.

```go
func (w *MachineWorker) Events() <-chan MachineEvent {
```

nvinuesa commented on internal/provisionertaskbis/machine_worker.go:
```diff
@@ -0,0 +1,598 @@
+	case EventZoneAssigned:
+		return w.handleZoneAssigned(ctx, event.Zone)
+	case EventZoneRequestFailed:
+		return w.handleZoneRequestFailed(ctx, event.ZoneError)
```
FSM handles this. Handlers check current state and ignore stale events.

nvinuesa committed:
chore: move provisionertaskbis to v2

Also, add another terminal state (`Error`) that will handle the retry
exhaustion.

nvinuesa commented on internal/provisionertaskbis/machine_worker.go:
```diff
@@ -0,0 +1,598 @@
+		return nil
+	}
+
+	// StartInstance succeeded.
```
Done.

nvinuesa committed:
chore: do not clear retryTimer when timer fires

nvinuesa commented on internal/provisionertaskbis/machine_worker.go:
```diff
@@ -0,0 +1,598 @@
+func (w *MachineWorker) retryTimerChan() <-chan time.Time {
+	if w.retryTimer == nil {
+		return nil
+	}
```
I've been trying to do this in an elegant way, but it's not simple. The issue is that we schedule retries on the handler* methods, so if we want the retryTimer to be local to the main `loop()`, then we would have to be passing (signals?) back and forth from the handlers, which makes it very difficult to maintain.
Also, the retry timer is only ever called from within the main loop method stack so it should be safe for concurrency.

Do you have an idea about how else to do this? Otherwise I'd continue working on this on follow-up patches which will include more workers if that's ok with you.

nvinuesa commented on internal/provisionertaskbis/machine_worker.go:
```diff
@@ -0,0 +1,598 @@
+	w.stopRetryTimer() // Cancel any existing timer.
+	w.retryTimer = time.NewTimer(w.config.RetryDelay)
+	w.config.Logger.Debugf(ctx, "machine %s scheduled retry in %v", w.config.MachineID, w.config.RetryDelay)
+}
```
Same as https://github.com/juju/juju/pull/21510#discussion_r2689613169.

manadart commented on internal/provisionertask/v2/events.go:
```diff
@@ -0,0 +1,117 @@
+	InstanceID string
+	ZoneName   string
+	Success    bool
+	Error      error // Only set if Success is false.
```
Why have both? Just use a nil-able error.

manadart reviewed (commented):
The direction is good here. We should pursue it.

Some question marks over the state transitions.

nvinuesa committed:
refactor(provisionertask): handle machine dying state and rename types for clarity

Distinguish dying from dead in the FSM life-change handler so the provisioner
no longer attempts to provision or de-provision a machine whose agent workers
are still shutting down.

nvinuesa commented on internal/provisionertask/v2/events.go:
```diff
@@ -0,0 +1,117 @@
+
+const (
+	// EventLifeChanged indicates the machine's life value changed.
+	EventLifeChanged MachineEventType = "LifeChanged"
```
Yeah sure I agree, see b6735504f954bb194b562ceb48a70ebdf6bd05d9.

nvinuesa commented on internal/provisionertask/v2/fsm.go:
```diff
@@ -0,0 +1,148 @@
+)
+
+// State represents the FSM state of a machine worker.
+type State string
```
Indeed, renamed to `MachineState`.

nvinuesa commented on internal/provisionertask/v2/machine_worker.go:
```diff
@@ -0,0 +1,597 @@
+			}
+			return w.doRemove(ctx)
+		}
+		// life.Alive - check if already provisioned.
```
Very good catch, thanks. Added a condition to check for `Dying` and in that case we don't provision but just wait for it to become Dead.

nvinuesa commented on internal/provisionertask/v2/machine_worker.go:
```diff
@@ -0,0 +1,597 @@
+				w.config.MachineID, w.instanceID)
+			return w.transitionTo(ctx, StateRunning)
+		}
+		// Not provisioned yet, start provisioning.
```
Right, answered in https://github.com/juju/juju/pull/21510#discussion_r2809506751.

nvinuesa commented on internal/provisionertask/v2/machine_worker.go:
```diff
@@ -0,0 +1,597 @@
+			}
+			return w.doRemove(ctx)
+		}
+		// Still alive, waiting for zone response.
```
Again, very good catch. Fixed also by checking for `Dying` and transition to Pending in that case.

nvinuesa commented on internal/provisionertask/v2/machine_worker.go:
```diff
@@ -0,0 +1,597 @@
+		// needed.
+
+	case StateRunning:
+		if newLife == life.Dead {
```
Added comments on `handleLifeChange`.

> do we need to look for the deletion mark (it's a proxy)...

I was thinking exactly about this when working on storage removal this week. Indeed I think the removal domain *should* be able to fully clean up everything and the provisioner task only do de-provisioning. Will work on this on future patches (for the moment I'm keeping them simple).

nvinuesa commented on internal/provisionertask/v2/machine_worker.go:
```diff
@@ -0,0 +1,597 @@
+	req := NewZoneRequest(w.config.MachineID, w.config.DistributionGroup, w.config.Constraints)
+
+	select {
+	case w.config.RequestChan <- req:
```
Removed RequestCancelZone, NewCancelZoneRequest, and the cancelZoneRequest method entirely. The stale zone response is already safely handled: when the zone assignment arrives after the worker has transitioned away from StateRequestingZone, handleZoneAssigned ignores it via the state check. No cancel message needed.

hpidcock commented on internal/provisionertask/v2/machine_worker.go:
```diff
@@ -0,0 +1,605 @@
+		if err := w.transitionTo(ctx, StateRemoving); err != nil {
+			return err
+		}
+		return w.doRemove(ctx)
```
Let the outer loop call the next state's do.

hpidcock reviewed (commented):
I think this needs to leverage the FSM more. There is a lot of `doX` that does a state transition then does more work. I would break each of these into discrete operations that can be tested on their own. The machine is then driven from the state, not just reflected in the state.

nvinuesa committed:
refactor(provisionertask): implement state-entry pattern and discrete
FSM states

Replace the monolithic doX() methods with a state-driven loop where each
active FSM state has exactly one discrete, testable entry action.

Add StateRegistering (SetInstanceInfo) and StateRollingBack (orphan
cleanup
via StopInstances) to make the StartInstance → register → rollback path
explicit in the FSM rather than buried inside doProvision().

The main loop now calls driveState() before blocking on the select. If
driveState() changes the FSM state, the loop continues immediately
without
entering the select, keeping life events safely buffered in EventsChan
while
active operations chain through state transitions.

Event handlers (applyLifeChange, applyZoneAssigned,
applyZoneRequestFailed,
applyRetryTimer) are now pure state-transition functions with no side
effects.
All work is performed by the enter*() action methods driven by
driveState().

sinanawad commented:
Closing due to inactivity.
